File: java/com/commoncoder/learn/dsa/tree/NAryTreeTraversals.java
package com.commoncoder.learn.dsa.tree;

import java.util.*;

public class NAryTreeTraversals {

    public int countNodes(NAryTreeNode root) {
        if (root == null) {
            return 0;
        }
        int count = 1;
        for (NAryTreeNode child : root.children) {
            count += countNodes(child);
        }
        return count;
    }

    static class TraversalStackNode {
        NAryTreeNode node;
        boolean areAllChildrenExplored;

        public TraversalStackNode(NAryTreeNode node) {
            this.node = node;
        }
    }

    // TC: O(N) SC: O(N)
    public List<Integer> postorder(NAryTreeNode root) {
        List<Integer> result = new ArrayList<>();
        if (root == null) {
            return result;
        }
        Stack<TraversalStackNode> stack = new Stack<>();
        stack.push(new TraversalStackNode(root));
        while (!stack.isEmpty()) {
            TraversalStackNode top = stack.peek();
            if (top.areAllChildrenExplored || (top.node.children.isEmpty())) {
                result.add(top.node.data);
                stack.pop();
                continue;
            }
            top.areAllChildrenExplored = true;
            for (int i = top.node.children.size() - 1; i >= 0; i--) {
                stack.push(new TraversalStackNode(top.node.children.get(i)));
            }
        }
        return result;
    }

    // N nodes in tree
    // TC: O(N) SC: O(N)
    public List<Integer> postorderRecursive(NAryTreeNode root) {
        List<Integer> result = new ArrayList<>();
        postorderRecursive(root, result);
        return result;
    }

    private void postorderRecursive(NAryTreeNode root, List<Integer> result) {
        if (root == null) {
            return;
        }
        for (NAryTreeNode child : root.children) {
            postorderRecursive(child, result);
        }
        result.add(root.data); // since this is postorder, we process parent last
    }

    // N nodes in tree
    // TC: O(N) SC: O(N)
    public List<Integer> preorderRecursive(NAryTreeNode root) {
        List<Integer> result = new ArrayList<>();
        preorderRecursive(root, result);
        return result;
    }

    public List<Integer> preorder(NAryTreeNode root) {
        List<Integer> result = new ArrayList<>();
        if (root == null) {
            return result;
        }
        Stack<NAryTreeNode> stack = new Stack<>();
        stack.push(root);
        while (!stack.isEmpty()) {
            NAryTreeNode top = stack.pop();
            result.add(top.data);
            for (int i = top.children.size() - 1; i >= 0; i--) {
                stack.push(top.children.get(i));
            }
        }
        return result;
    }

    private void preorderRecursive(NAryTreeNode root, List<Integer> result) {
        if (root == null) {
            return;
        }
        result.add(root.data); // since this is preorder, we process parent first
        for (NAryTreeNode child : root.children) {
            preorderRecursive(child, result);
        }
    }

    // N nodes in tree
    // TC: O(N) SC: O(N)
    private boolean dfs(NAryTreeNode root, int valToSearch) {
        if (root == null) {
            return false;
        }
        if (root.data == valToSearch) {
            return true;
        }
        for (NAryTreeNode child : root.children) {
            boolean result = dfs(child, valToSearch);
            if (result) {
                return true;
            }
        }
        return false;
    }

    // TC: O(N) SC: O(N)
    public List<Integer> levelOrderTraversal(NAryTreeNode root) {
        List<Integer> result = new ArrayList<>();
        if (root == null) {
            return result;
        }
        Queue<NAryTreeNode> queue = new LinkedList<>();
        queue.add(root);
        while (!queue.isEmpty()) {
            NAryTreeNode front = queue.poll();
            result.add(front.data);
            for (NAryTreeNode child : front.children) {
                queue.add(child);
            }
        }
        return result;
    }

    // TC: O(N) SC: O(N)
    public boolean bfs(NAryTreeNode root, int valToSearch) {
        if (root == null) {
            return false;
        }
        Queue<NAryTreeNode> queue = new LinkedList<>();
        queue.add(root);
        while (!queue.isEmpty()) {
            NAryTreeNode front = queue.poll();
            if (front.data == valToSearch) {
                return true;
            }
            for (NAryTreeNode child : front.children) {
                queue.add(child);
            }
        }
        return false;
    }

    public static void main(String[] args) {
        NAryTreeNode five = new NAryTreeNode(5);
        NAryTreeNode six = new NAryTreeNode(6);
        NAryTreeNode seven = new NAryTreeNode(7);
        NAryTreeNode eight = new NAryTreeNode(8);
        NAryTreeNode nine = new NAryTreeNode(9);
        NAryTreeNode ten = new NAryTreeNode(10);

        NAryTreeNode two = new NAryTreeNode(2, Arrays.asList(five, six));
        NAryTreeNode three = new NAryTreeNode(3, Arrays.asList(seven));
        NAryTreeNode four = new NAryTreeNode(4, Arrays.asList(eight, nine, ten));
        NAryTreeNode root = new NAryTreeNode(1, Arrays.asList(two, three, four));

        NAryTreeTraversals NAryTreeTraversals = new NAryTreeTraversals();
        System.out.println(NAryTreeTraversals.postorderRecursive(root));
        System.out.println(NAryTreeTraversals.postorder(root));
    }

}

File: cpp/32TreeBasics/nary_tree_traversals.cpp
#include <bits/stdc++.h>

using namespace std;

class NAryTreeNode {
 public:
  int data;
  vector<NAryTreeNode*> children;

  NAryTreeNode(int data) { this->data = data; }

  NAryTreeNode(int data, vector<NAryTreeNode*> children) {
    this->data = data;
    this->children = children;
  }
};

// TC: O(N) SC: O(N)
vector<int>* preorder(NAryTreeNode* root) {
  vector<int>* result = new vector<int>();
  if (root == nullptr) {
    return result;
  }
  stack<NAryTreeNode*> s;
  s.push(root);
  while (!s.empty()) {
    NAryTreeNode* topVal = s.top();
    s.pop();
    result->push_back(topVal->data);
    for (auto i = topVal->children.size() - 1; i >= 0; i--) {
      s.push(topVal->children[i]);
    }
  }
  return result;
}

class TraversalStackNode {
 public:
  NAryTreeNode* node;
  bool isAllPreprcoessingDone;

  TraversalStackNode(NAryTreeNode* node) {
    this->node = node;
    this->isAllPreprcoessingDone = false;
  }
};

// TC: O(N) SC: O(N)
vector<int>* postorder(NAryTreeNode* root) {
  vector<int>* result = new vector<int>();
  if (root == nullptr) {
    return result;
  }
  stack<TraversalStackNode*> s;
  s.push(new TraversalStackNode(root));
  while (!s.empty()) {
    TraversalStackNode* top = s.top();
    if (top->isAllPreprcoessingDone || top->node->children.empty()) {
      result->push_back(top->node->data);
      s.pop();
      continue;
    }
    top->isAllPreprcoessingDone = true;
    for (int i = top->node->children.size() - 1; i >= 0; i++) {
      s.push(new TraversalStackNode(top->node->children[i]));
    }
  }
  return result;
}

// TC: O(N) SC: O(N)
vector<int>* preorderRecursive(NAryTreeNode* root) {
  vector<int>* result = new vector<int>();
  preorderRecursive(root, result);
  return result;
}

void preorderRecursive(NAryTreeNode* root, vector<int>* result) {
  if (root == nullptr) {
    return;
  }
  result->push_back(root->data);
  for (NAryTreeNode* child : root->children) {
    preorderRecursive(child, result);
  }
}

NAryTreeNode* dfs(NAryTreeNode* root, int valToSearch) {
  if (root == nullptr) {
    return nullptr;
  }
  if (root->data == valToSearch) {
    return root;
  }
  for (NAryTreeNode* child : root->children) {
    NAryTreeNode* searchResult = dfs(child, valToSearch);
    if (searchResult != nullptr) {
      return searchResult;
    }
  }
  return nullptr;
}

int countNodes(NAryTreeNode* root) {
  if (root == nullptr) {
    return 0;
  }
  int count = 1;
  for (NAryTreeNode* child : root->children) {
    count += countNodes(child);
  }
  return count;
}

// TC: O(N) SC: O(N)
vector<int>* postorderRecursive(NAryTreeNode* root) {
  vector<int>* result = new vector<int>();
  postorderRecursive(root, result);
  return result;
}

void postorderRecursive(NAryTreeNode* root, vector<int>* result) {
  if (root == nullptr) {
    return;
  }
  for (NAryTreeNode* child : root->children) {
    postorderRecursive(child, result);
  }
  result->push_back(root->data);
}

// TC: O(N) SC: O(N)
vector<int>* levelOrderTraversal(NAryTreeNode* root) {
  vector<int>* result = new vector<int>();
  if (root == nullptr) {
    return result;
  }
  queue<NAryTreeNode*> q;
  q.push(root);
  while (!q.empty()) {
    NAryTreeNode* firstNode = q.front();
    q.pop();

    result->push_back(firstNode->data);

    for (NAryTreeNode* child : root->children) {
      q.push(child);
    }
  }

  return result;
}

vector<vector<int>>* levelOrderTraversalSplittedByLevels(NAryTreeNode* root) {
  vector<vector<int>>* result = new vector<vector<int>>();
  if (root == nullptr) {
    return result;
  }
  queue<NAryTreeNode*> q;
  q.push(root);

  while (!q.empty()) {
    vector<int> levelAns;
    int nextLevelSize = 0;
    int levelSize = q.size();
    while (levelSize > 0) {
      NAryTreeNode* firstNode = q.front();
      q.pop();

      levelAns.push_back(firstNode->data);

      for (NAryTreeNode* child : root->children) {
        q.push(child);
      }
      levelSize--;
    }
    result->push_back(levelAns);
  }
}

vector<vector<int>>* zigzagTraversal(NAryTreeNode* root) {
  vector<vector<int>>* result = new vector<vector<int>>();
  if (root == nullptr) {
    return result;
  }
  queue<NAryTreeNode*> q;
  q.push(root);

  while (!q.empty()) {
    vector<int> levelAns;
    int nextLevelSize = 0;
    int levelSize = q.size();
    int levelCount = 0;
    while (levelSize > 0) {
      NAryTreeNode* firstNode = q.front();
      q.pop();

      levelAns.push_back(firstNode->data);

      for (NAryTreeNode* child : root->children) {
        q.push(child);
      }
      levelSize--;
    }
    if (levelCount % 2 == 1) {
      levelAns.reserve(levelAns.size());
    }
    result->push_back(levelAns);
    levelCount++;
  }
}

NAryTreeNode* bfs(NAryTreeNode* root, int valToSearch) {
  if (root == nullptr) {
    return nullptr;
  }
  queue<NAryTreeNode*> q;
  q.push(root);
  while (!q.empty()) {
    NAryTreeNode* firstNode = q.front();
    q.pop();

    if (firstNode->data == valToSearch) {
      return firstNode;
    }

    for (NAryTreeNode* child : root->children) {
      q.push(child);
    }
  }

  return nullptr;
}
